File: 347-top-k-frequent-elements/347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        n = {}
        counts =[[] for i in range(len(nums) + 1)]
        
        for num in nums:
            if num in n:
                n[num] += 1
            else:
                n[num] = 1
        
        for num, count in n.items():
            counts[count].append(num)
        
        res = []
        
        for i in range(len(counts) - 1, -1, -1):
            for num in counts[i]:
                res.append(num)
                if len(res) == k:
                    return res
                

File: 347-top-k-frequent-elements/347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */

var topKFrequent = function(nums, k) {
    const res = [];
    const n = {};
    
    for (let num of nums) {
        if (num in n) n[num]++
        else n[num] = 1
    }
    
    // counts indices will rep count of that num
    const counts = [];
    for (let i = 0; i <= nums.length; i++) {
        let r = [];
        counts.push(r);
    }
    for (let num in n) {
        counts[n[num]].push(num);
    }

    let index = counts.length - 1;
    while (k > 0) {
        if (counts[index].length) {
            for (let i = 0; i < counts[index].length; i++) {
                if (k === 0) break;
                res.push(counts[index][i]);
                k--;
            }
        }
        index--;
    }
    return res;
};